vowpalwabbit/parse_example.cc

```diff
@@ -76,9 +76,15 @@ class TC_parser
 
   inline void parserWarning(const char* message, char* begin, char* pos, const char* message2)
   {
-    cerr << message << std::string(begin, pos - begin).c_str() << message2 << "in Example #"
-         << this->p->end_parsed_examples << ": \"" << std::string(this->beginLine, this->endLine).c_str() << "\""
-         << endl;
+    std::stringstream ss;
+    ss << message << std::string(begin, pos - begin).c_str() << message2 << "in Example #"
+       << this->p->end_parsed_examples << ": \"" << std::string(this->beginLine, this->endLine).c_str() << "\""
+       << endl;
+    if (p->strict_parse) {
+      throw VW::strict_parse_exception(ss.str());
```

Comment:
A couple of concerns with this behavior.

1) There will be two substantially different code paths if exceptions/errors are enabled vs not.  While this was probably an acceptable tradeoff for memory allocation issues, this is not desireable in general.  For example we cannot recover and continue processing examples.

A couple of ways to avoid this issue:
a) Always throw.  Stop parsing the current example.  Let the outer loop either terminate or recover.  This would require some modifications so that the parser can continue parsing.
b) Always return error.  Stop parsing the current example.  Let the outer loop terminate or continue.
c) Accumulate the errors and continue parsing.  Continue or terminate in the outer loop based on a command line flag

2) If it is desireable to recover and continue parsing, then the performance impact of throwing compared to error codes is substantial.  (For example: https://github.com/rajan-chari/perf_tests) In high density type applications, this would require substantial CPU overallocation to be safe.  

I get that this might fall outside the scope of this PR.

Reply:
I'm not entirely clear on (1) in this case.  In particular, adding an exception in the use case that @palmerlao has in mind is a slightly more verbose form of exit() in my understanding.

If we had a global "terminate on error" option for VW, then none of the performance impact is relevant?  (The goal of such an option is fast debugging of a data pipeline.)

Reply:
Clarifying (1)
As an application, terminate_on_error() makes sense.  As a library, I am not so sure.  It would be good to have continue_on_error() for example.  The exception leaves the parser in an undefined state.  It would be better to have some mechanism for letting the error bubble up.  exit_on_error() or continue_on_error() can be decided higher up the stack.

Reply:
I agree. 

The present default behavior is warn-and-continue.  This adds an option for terminate_on_error().  

In the longer term, having the warnings bubbled up for a later decision does seem better.  Probably beyond the scope of this pull request though?

Reply:
Yes makes sense.  It's a larger change.  We would likely need this for atomization.  I would like to revisit this usage of exceptions at that point if there are no objections.  I will add an issue to track this.